### ISSUE TITLE
1.5 performance regressions

### ISSUE THREAD

**Issue by colinsurprenant** [elastic/logstash#2870]
This is a master issue to track 1.5 performance regression and proposed fixes.

I will add measurements between 1.4.2, 1.5.0-rc2 and the fixes in #2869

relates to #2859

**Comment by colinsurprenant** [comment]
simple output stage tests, with improvements for the output function generation in `fix/perf_regression`

- 1.4.2
```sh
USE_RUBY=1 bin/logstash --quiet -e 'input { generator { count => 10000000 } } output { stdout { codec => dots } }' | pv -Wbart > /dev/null
9.54MiB 0:01:33 [ 104kiB/s] [ 104kiB/s]
```
- 1.5.0 master (1.5.0-rc2)
```sh
USE_RUBY=1 bin/logstash --quiet -e 'input { generator { count => 10000000 } } output { stdout { codec => dots } }' | pv -Wbart > /dev/null
9.54MiB 0:01:49 [89.5kiB/s] [89.5kiB/s]
```
- 1.5.0 fix/perf_regression
```sh
USE_RUBY=1 bin/logstash --quiet -e 'input { generator { count => 10000000 } } output { stdout { codec => dots } }' | pv -Wbart > /dev/null
9.54MiB 0:01:34 [ 103kiB/s] [ 103kiB/s]
```

**Comment by colinsurprenant** [comment]
- output stage with conditionals

#### 1.4.2
```sh
USE_RUBY=1 bin/logstash --quiet -e 'input { generator { count => 10000000 } } output { if [foo] == 1 {stdout{codec => dots}} if [bar] == 1 {stdout{codec => dots}} if [baz] == 1 {stdout{codec => dots}} stdout{codec => dots}}' | pv -Wbart > /dev/null
9.54MiB 0:02:40 [60.9kiB/s] [60.9kiB/s]
```

#### 1.5.0 master (1.5.0-rc2)
```sh
USE_RUBY=1 bin/logstash --quiet -e 'input { generator { count => 10000000 } } output { if [foo] == 1 {stdout{codec => dots}} if [bar] == 1 {stdout{codec => dots}} if [baz] == 1 {stdout{codec => dots}} stdout{codec => dots}}' | pv -Wbart > /dev/null
9.54MiB 0:02:23 [  68kiB/s] [  68kiB/s]
```

#### 1.5.0 fix/perf_regression
```sh
USE_RUBY=1 bin/logstash --quiet -e 'input { generator { count => 10000000 } } output { if [foo] == 1 {stdout{codec => dots}} if [bar] == 1 {stdout{codec => dots}} if [baz] == 1 {stdout{codec => dots}} stdout{codec => dots}}' | pv -Wbart > /dev/null
9.54MiB 0:02:03 [  79kiB/s] [  79kiB/s]
```

**Comment by colinsurprenant** [comment]
- simple no-op filter stage 
- note that at this point there is no improvements committed for the filter stage in fix/perf_regression
- we can see the filter stage perf regression between 1.4.2 and 1.5.0-rc2. the slight improvement in fix/perf_regression is due only to the output stage fixes

```sh
USE_RUBY=1 bin/logstash --quiet -e 'input{generator{count => 10000000 }} filter{clone{}} output{stdout{codec => dots}}' | pv -Wbart > /dev/null
```

Version | Rate
---- | ----
1.4.2 | 9.54MiB 0:02:23 [68.3kiB/s] [68.3kiB/s]
1.5.0 master (1.5.0-rc2) | 9.54MiB 0:02:43 [59.9kiB/s] [59.9kiB/s]
1.5.0 fix/perf_regression | 9.54MiB 0:02:33 [63.5kiB/s] [63.5kiB/s]

**Comment by colinsurprenant** [comment]
#### filter stage with conditionals 
- note that at this point there is no improvements committed for the filter stage in fix/perf_regression

```sh
USE_RUBY=1 bin/logstash --quiet -e 'input{generator{count => 10000000 }} filter{if [foo] == 1 {clone{}} if [bar] == 1 {clone{}} if [baz] == 1 {clone{}} clone{}} output{stdout{codec => dots}}' | pv -Wbart > /dev/null
```

Version | Rate
---- | ----
1.4.2 | 9.54MiB 0:03:22 [48.1kiB/s] [48.1kiB/s]
1.5.0 master (1.5.0-rc2) | 9.54MiB 0:03:22 [48.2kiB/s] [48.2kiB/s]
1.5.0 fix/perf_regression | 9.54MiB 0:03:21 [48.3kiB/s] [48.3kiB/s]

**Comment by colinsurprenant** [comment]
So I will be spending a bit of time on trying to improve the filter stage code generation.

**Comment by colinsurprenant** [comment]
nice improvements in last commit in #2869, will update benchmarks shortly!

**Comment by grecco-gn** [comment]
I just checkout the latest commit from fix/perf_regression branch but on my instance (AWS/EC2 c3.2xlarge, Linux Ubuntu LTS 14.04, Oracle Java 1.7.0_67, vendored Jruby) I still get much worse result compared 1.4.2

I'm benchmarking using the "simple output stage" command.
Using a vanilla 1.4.2 (downloaded from logstash.net) I get:  9.54MB 0:01:41 [96.1kB/s] [96.1kB/s]
On 1.5 branch fix/perf_regression: 9.54MB 0:04:35 [35.4kB/s] [35.4kB/s]B/s]

I tried to profile them both version using VisualVM and on 1.5 I noticed that logstash is spending an awful amount of time on LogStash::Event init_timestamp, so I commented out line 70 of lib/logstash/event.rb 
```ruby
#@data[TIMESTAMP] = init_timestamp(@data[TIMESTAMP])
```
and I added the following code (slightly adapted) from 1.4.2 

```ruby
    if @data.include?(TIMESTAMP)
      t = @data[TIMESTAMP]
      if t.is_a?(String)
        @data[TIMESTAMP] = LogStash::Time.parse_iso8601(t)
      end
    else
      @data[TIMESTAMP] = ::Time.now.utc
    end
```

The result for this modified version are: 9.54MB 0:03:03 [53.4kB/s] [53.4kB/s]]
which is not yet at 1.4.2 level but still much better than earlier.

So, this one clearly is not a fix but maybe it is worth investigating by someone more experienced in the Logstash inners.

**Comment by colinsurprenant** [comment]
@grecco-gn thanks for the report and profiling hint, I will followup on this today! could you make sure both tests are performed using the same JVM options, see my last comment in #2859

**Comment by grecco-gn** [comment]
Yes, I tried multiple combinations:
 - Vendored Jruby
 - Local jruby with USE_RUBY=1
 - With and without "-XX:+UseConcMarkSweepGC"
 - I even tried using the 1.4.2 startup scripts to start 1.5
but I could not see much changes among the different options. 

Consider also that, for some reasons, on the EC2 instance results fluctuate a lot (+/- 5K/s) so it is hard to measure small changes.

**Comment by colinsurprenant** [comment]
@grecco-gn I added a new `Event` optimization commit 99484fc1397b81c468abe7a32fb03a71b7cb8a33 in #2869 for the timestamp initialization and also realized the `@logger` set using `Cabin::Channel#get` call was costly by going through mutex synchronization and was totally not required at `Event` initialization.

I will update all performance numbers now.

**Comment by colinsurprenant** [comment]
I just updated all performance numbers.

**Comment by colinsurprenant** [comment]
@grecco-gn did you get a chance to test with the latest changes?

**Comment by grecco-gn** [comment]
Yes, thank you @colinsurprenant, I did it juts now. I tried it multiple times and on average I get around 65K/s but it peaks often up to 75K/s. 
It is way better than earlier (even if not at 1.4.2 level yet, where on the same server it averages around 95K/s).

Profiling doesn't show up any particularly suspicious method now, but I noticed this:

If I do "top -H", which reports on linux threads, i can see the generator thread taking 99.9% of the CPU and the output thread taking about 91% which sort of makes sense considering that creating an event is probably more intensive than what codec-dots does.

However the profiler reports something possibly in contrast with that. 
Essentially it seems that a lot of time (like 60 seconds out of the about 3 minutes to process 10,000,000 events) is burnt on SizedQueue.push while only about 15 seconds reading from the queue, and this is what you would expect if the consumer would be slower than the generator.

I haven't investigated much further that that, but maybe this already suggest you something.

Thank you for your support!.
